ledger/internal/apptxn_test.go

```diff
@@ -2439,3 +2611,195 @@ func BenchmarkMaximumCallStackDepth(b *testing.B) {
 		executeMegaContract(b)
 	}
 }
+
+// TestInnerClearState ensures inner ClearState performs close out properly, even if rejects.
+func TestInnerClearState(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	l := newTestLedger(t, genBalances)
+	defer l.Close()
+
+	// inner will be an app that we opt into, then clearstate
+	// note that clearstate rejects
+	inner := txntest.Txn{
+		Type:              "appl",
+		Sender:            addrs[0],
+		ApprovalProgram:   "int 1",
+		ClearStateProgram: "int 0",
+		LocalStateSchema: basics.StateSchema{
+			NumUint:      2,
+			NumByteSlice: 2,
+		},
+	}
+
+	eval := nextBlock(t, l, true, nil)
+	txn(t, l, eval, &inner)
+	vb := endBlock(t, l, eval)
+	innerId := vb.Block().Payset[0].ApplicationID
+
+	outer := txntest.Txn{
+		Type:   "appl",
+		Sender: addrs[0],
+		ApprovalProgram: main(`
+itxn_begin
+ int appl
+ itxn_field TypeEnum
+ txn Applications 1
+ itxn_field ApplicationID
+itxn_submit
+`),
+		ForeignApps: []basics.AppIndex{innerId},
+	}
+
+	eval = nextBlock(t, l, true, nil)
+	txn(t, l, eval, &outer)
+	vb = endBlock(t, l, eval)
+	outerId := vb.Block().Payset[0].ApplicationID
```

Comment:
What's the purpose of the outer app in this test? It doesn't seem used in a relevant way

Reply:
The intent is just to use outer as an app that opts its account into inner, then clears state with an inner, to make sure inner clears work. I'm on phone now, so I'll confirm that's actually what's going on when I get back to computer.

Reply:
I got it wrong, I was doing the doing test of a normal account opting into inner.

Reply:
No it tests inner clear state.  I'll get the budget stuff a bit later.